log/logging.go

```diff
@@ -188,22 +201,118 @@ func SetTimestampFlags(flags int) {
 	DefaultLogger.SetTimestampFlags(flags)
 }
 
+type LoggerInterface interface {
+	Fatal(description string, keysAndValues ...interface{})
+	Error(description string, keysAndValues ...interface{})
+	Warn(description string, keysAndValues ...interface{})
+	Info(description string, keysAndValues ...interface{})
+	Debug(description string, keysAndValues ...interface{})
+	Trace(description string, keysAndValues ...interface{})
+	SetOutput(w io.Writer)
+	SetTimestampFlags(flags int)
+	SetStaticField(name string, value interface{})
+}
+
+func NewLogger(format LogFormat, id string, staticKeysAndValues ...interface{}) LoggerInterface {
+	return newLoggerStruct(format, id, staticKeysAndValues...)
+}
+
+func newLoggerStruct(format LogFormat, id string, staticKeysAndValues ...interface{}) *Logger {
+	var prefix string
+	var flags int
+	var formatter logFormatter
+	staticArgs := make(map[string]string, 0)
+
+	format = SanitizeFormat(format)
+
+	if format == JsonFormat {
+		formatter = jsonLogFormatter{}
+
+		// Don't mess up the json by letting logger print these:
+		prefix = ""
+		flags = 0
+
+		// Instead put them into the staticArgs
+		if defaultPrefix != "" {
+			staticArgs["prefix"] = defaultPrefix
+		}
+	} else {
+		formatter = plainTextLogFormatter{}
+		prefix = defaultPrefix
+		flags = Flags
+	}
+
+	// Do this after handling prefix, so that individual loggers can override
+	// external env variable.
+	currentKey := ""
+	for i, arg := range staticKeysAndValues {
+		if i%2 == 0 {
+			currentKey = fmt.Sprintf("%v", arg)
+		} else {
+			staticArgs[currentKey] = fmt.Sprintf("%v", arg)
+		}
+	}
+
+	// If there are an odd number of keys+values, add the dangling key with empty
+	// value.
+	if len(staticKeysAndValues)%2 == 1 {
+		staticArgs[currentKey] = ""
+	}
+
+	return &Logger{
+		ID:    id,
+		Level: Level,
+
+		formatter:  formatter,
+		staticArgs: staticArgs,
+
+		// don't touch the default logger on 'log' package
+		l: log.New(defaultOutput, prefix, flags),
+	}
+}
+
+type LogFormat string
+
+const (
+	DefaultFormat   LogFormat = "" // Use env variable, defaulting to PlainTextFormat
+	PlainTextFormat           = "text"
+	JsonFormat                = "json"
+)
+
+func SanitizeFormat(format LogFormat) LogFormat {
+	if format == PlainTextFormat || format == JsonFormat {
+		return format
+	} else {
+		// Whether it's explicitly a DefaultFormat, or it's an unrecognized value,
+		// try to take from env var.
+		envFormat := os.Getenv("DEFAULT_LOG_ENCODING_FORMAT")
```

Comment:
Actually, since there won't be multiple logging packages per project (right? :badpokerface:) keeping this un-namedspaced is totally fine. In this case I'd maybe just swap (or even drop) "DEFAULT" and "LOG" to keep all env vars consistently prefixed as "LOG_": "LOG_ENCODING_FORMAT"?